library/HTMLPurifier/AttrDef/CSS/Gradient.php

```diff
@@ -0,0 +1,153 @@
+
+        $linear = false;
+        if (strpos($function, 'linear') !== false) {
+            $linear = true;
+        }
+
+        preg_match('#gradient\((.*)\)#', $string, $values);
+        $values = end($values);
+
+        if (substr_count($values, '(') !== substr_count($values, ')')) {
+            return false;
+        }
+
+        $values = (explode(',', $values));
+        $bracket = false;
+        $parts = [];
+
+        for ($i = 0; $i < count($values); $i++) {
+            if (strpos($values[$i], '(') !== false) {
+                $bracket = $values[$i];
+            } else if ($bracket !== false) {
+                $bracket .= ',' . trim($values[$i]);
+
+                if(strpos($values[$i], ')') !== false) {
+                    $parts[] = trim($bracket);
+                    $bracket = false;
+                }
+            } else {
+                $parts[] = trim($values[$i]);
+            }
+        }
+
+        $final = '';
+        $i = 0;
+        foreach ($parts as $part) {
+            $i++;
+            if (ctype_space($part)) {
+                continue;
+            }
+
+            // test for direction or angle but only for the first parameter of linear function
+            if ($i === 1 && $linear === true) {
+                $r = $this->direction->validate($part, $config, $context);
+                if ($r !== false) {
+                    $final .= $r.',';
+                    continue;
+                }
+
+                $r = $this->angle->validate($part, $config, $context);
+                if ($r !== false) {
+                    $final .= $r.',';
+                    continue;
+                }
+            }
+
+            if ($i === 1 && $linear === false) {
+                if (in_array(trim($part), ['circle', 'ellipse'], true)) {
+                    $final .= trim($part).',';
+                    continue;
+                }
+            }
+
+            // test for color
+            $size = false;
+            if (preg_match('#(.*)\s+(\d+(\.\d+)?(%|px|ex|em|in|cm|mm|pt|pc))\s*$#', $part, $matches)) {
+                $color = $matches[1];
+                $size = $matches[2];
+            } else {
+                $color = $part;
+            }
+
+            $r = $this->color->validate($color, $config, $context);
+            if ($r !== false) {
+                $final .= $r;
+
+                if ($size !== false) {
+                    $r = $this->size->validate($size, $config, $context);
+                    if ($r !== false) {
+                        $final .= ' '.$r;
+                    }
+                }
+                $final .= ',';
+            }
+
+            // the different size keyword is not yet implemented
+            // correct values are: closest-side, farthest-side, closest-corner, farthest-corner
+
+        }
+
+        if ($final === '') {
+            return false;
+        }
+
+        return $function . '(' . rtrim($final, ', ') . ')';
```

Comment:
It doesn't look like the value of function was ever validated. There is only a fixed number of allowed ones, right?